### PR TITLE
Use less stack space to parse if statements with attributes

### PR DESCRIPTION
**Pull request by RikkiGibson**
This PR is basically exploratory--trying to figure out what it would take to fix the stack depth issue discussed in #41240 without reverting the attributes change. There are some temporary tweaks such as commenting out catch blocks which would be reverted before merge.

The nature of the change is to inline more things (particularly on x64) and to pull more locals off the stack (in the case of x86). At this time I am leaning toward reverting the statement attributes change in #38937, and possibly taking it back in at a future point, after the feature has been merged. But I would like to get input on this. I will open another PR to explore a revert strategy while still having decent diagnostics on incomplete code (i.e. an attribute list in the middle of a method, or an attribute list on an incomplete local function).

Here are some native stack measurements based on the scenario in EndToEndTest.NestedIfStatements:

### csc master branch, x64

Bytes | Method
-- | --
**640.00** | **Total**
160.00 | LanguageParser.ParseIfStatement+0x106
80.00 | LanguageParser.ParseStatementCore+0xa2
112.00 | LanguageParser.ParseStatements+0x6c
112.00 | LanguageParser.ParseBlock+0x15e
80.00 | LanguageParser.ParseStatementCore+0xa2
96.00 | LanguageParser.ParseEmbeddedStatement+0x17

### csc features/local-function-attributes branch, x64

Bytes | Method
-- | --
**928.00** | **Total**
176.00 | LanguageParser.ParseIfStatement(SyntaxList)
80.00 | LanguageParser.TryParseStatementNoDeclaration(SyntaxList\`1, Boolean)
128.00 | LanguageParser.TryParseStatementCore()
112.00 | LanguageParser.ParseStatements(CSharpSyntaxNode ByRef, SyntaxListBuilder\`1, Boolean)
128.00 | LanguageParser.ParseBlock(SyntaxList\`1, Boolean, Boolean)
80.00 | LanguageParser.TryParseStatementNoDeclaration(SyntaxList\`1, Boolean)
128.00 | LanguageParser.TryParseStatementCore()
96.00 | LanguageParser.ParseEmbeddedStatement()

## csc lfa branch after fix ba602a6, x64
Bytes | Method
-- | --
**608.00** | **Total**
160.00 | LanguageParser.ParseIfStatement+0x103
144.00 | LanguageParser.TryParseStatementCore+0xe1
160.00 | LanguageParser.ParseBlock+0x1ed
144.00 | LanguageParser.TryParseStatementCore+0xe1

(This measurement is probably an *overestimate* now, as more progress was made in d5d7007)

---

### csc master branch, x86

Bytes | Method
-- | --
**276.00** | **Total**
40.00 | LanguageParser.ParseIfStatement+0xca
20.00 | LanguageParser.ParseStatementNoDeclaration+0x10b
44.00 | LanguageParser.ParseStatementCore+0x9f
40.00 | LanguageParser.ParseStatements+0x55
68.00 | LanguageParser.ParseBlock+0x121
20.00 | LanguageParser.ParseStatementNoDeclaration+0x1ad
44.00 | LanguageParser.ParseStatementCore+0x9f
36.00 | LanguageParser.ParseEmbeddedStatement+0x12

### csc lfa branch, x86

Bytes | Method
-- | --
**372.00** | **Total**
56.00 | LanguageParser.ParseIfStatement+0xce
20.00 | LanguageParser.TryParseStatementNoDeclaration+0xb4
80.00 | LanguageParser.TryParseStatementCore+0xd7
116.00 | LanguageParser.ParseBlock+0x19b
20.00 | LanguageParser.TryParseStatementNoDeclaration+0xf3
80.00 | LanguageParser.TryParseStatementCore+0xd7

## csc lfa branch after fix d5d7007, x86
Note that the size of the stack cycle went up compared to master, but we still can parse at the expected complexity in EndToEndTests.

Bytes | Method
-- | --
**312.00** | **Total**
44.00 | LanguageParser.ParseIfStatement+0x102
20.00 | LanguageParser.TryParseStatementNoDeclaration+0xb4
56.00 | LanguageParser.TryParseStatementCore+0xd7
116.00 | LanguageParser.ParseBlock+0x19b
20.00 | LanguageParser.TryParseStatementNoDeclaration+0xf3
56.00 | LanguageParser.TryParseStatementCore+0xd7

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -2095,7 +2098,7 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatementCore(SyntaxKind
                     _termState |= TerminatorState.IsPossibleStatementStartOrStop; // partial statements can abort if a new statement starts
 
                     // Any expression is allowed, not just expression statements:
-                    var statement = this.TryParseStatementNoDeclaration(attributes, allowAnyExpression: true);
+                    var statement = this.TryParseStatementNoDeclaration((attributes, allowAnyExpression: true));
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This looks absolutely ridiculous but it makes the difference between getting inlined or not on netframework x64.

**Comment by CyrusNajmabadi** [comment]
> At this time I am leaning toward reverting the statement attributes change in #38937, and possibly taking it back in at a future point, 

:(

Were the suggestions i made in https://github.com/dotnet/roslyn/pull/41240 not suitable?  i.e. why do we need so many indirect stack frames to do parsing of statements?  it seems unnecessary. We could likely get bcak a decent win by merging some of these forwarding methods right?

**Comment by CyrusNajmabadi** [comment]
> while still having decent diagnostics on incomplete code (i.e. an attribute list in the middle of a method, or an attribute list on an incomplete local function).

Note: the scenario is not just about decent diagnostics.  It's about actually representing these constructs so that the entire scope of IDE features can work well on an appropriate tree.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: yowza.  perhaps file bug on them?

**Comment by RikkiGibson** [comment]
Bleh, looks like this change has broken stack depth tests in other scenarios. May have to investigate.

> why do we need so many indirect stack frames to do parsing of statements? it seems unnecessary. We could likely get bcak a decent win by merging some of these forwarding methods right?

More effective inlining basically got us there on x64 which you can see in the measurements--8 stack frames in the feature branch, and just 4 stack frames in the fix branch. But on x86 I couldn't find an alternative to just pushing more locals off of the stack, for example with IfStatementBuilder. I think the netframework x86 JIT is less interested in inlining methods than x64 for whatever reason.

Were you thinking of maybe "de-abstracting" in the source code and actually getting rid of some intermediate methods?

Tagging @cston @davidwrighton whose help was invaluable in even getting me this far in this work.

**Comment by CyrusNajmabadi** [comment]
Looking at the parser, it really feels weird what we're doing with statement parsing. we've broken out all sorts of helpers that then end up reentering themselves multiple times for error-recovery and the like.

Part of the problem appears to eb in ParsePossibleDeclarationOrBadAwaitStatement where it tries three different approaches:

1. try to parse a declaration.
2. if we got parse errors, and we started with `await` try to parse as an `await expression`.
3. if we *still* got parse errors, go back to parsing as a declaration.

I posit that '3' just isn't an interesting case.  If we have parse errors, and the line starts with `await`, it's totally reasonable to just parse it out as an await expression, and still accept that, even if it has some diagnostics (which are likely to be something else entirely).  It's much more likely than the user having a type called 'await' which they are then writing code with.

Note that we would still parse legal code correctly in all cases.  it would just be reducing the amount of unnecessary recovery break-out logic for the extremely unlikely error recovery case.

**Comment by CyrusNajmabadi** [comment]
Looking into this now.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -6419,123 +6422,134 @@ private StatementSyntax TryParsePossibleDeclarationOrBadAwaitStatement(SyntaxLis
...
-            {
-                return this.ParseExpressionStatement(attributes);
-            }
+                if (this.IsPossibleLocalDeclarationStatement(allowAnyExpression))
+                {
+                    return null;
+                }
+                else
+                {
+                    return this.ParseExpressionStatement(default);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: this is a bug, forgot to change `default` back to `attributes` here.

**Comment by RikkiGibson** [comment]
Let me know if you learn anything, I would be curious to find out if more size reduction is possible by removing/simplifying certain error recovery paths.

**Comment by CyrusNajmabadi** [comment]
> Were you thinking of maybe "de-abstracting" in the source code and actually getting rid of some intermediate methods?

Yes.  For example, i don't see the need to have:

ParseStatementCore, ParseStatementNoDeclaration, and ParsePossibleDeclarationOrBadAwaitStatement.  I'm working on a set of PRs to demonstrate a simpler way to do things (which also should improve things like error recovery).  Can hopefully break it into small pieces to make it easier to review and get in.

**Comment by RikkiGibson** [comment]
Cool, please try testing your change by enabling the EndToEndTests.NestedIfStatement scenario, and let me know if you need any help measuring the native stack frame sizes. I imagine this might not be your first rodeo, but just in case.

**Comment by CyrusNajmabadi** [comment]
Will do toniht!

**Pull request closed without merging**